File: src/demos/vehicle/terrain/demo_VEH_CRMTerrain_WheeledVehicle.cpp
// =============================================================================
// PROJECT CHRONO - http://projectchrono.org
//
// Copyright (c) 2023 projectchrono.org
// All rights reserved.
//
// Use of this source code is governed by a BSD-style license that can be found
// in the LICENSE file at the top level of the distribution and at
// http://projectchrono.org/license-chrono.txt.
//
// =============================================================================
// Author: Radu Serban
// =============================================================================
//
// Polaris wheeled vehicle on CRM terrain
//
// =============================================================================

#include <cstdio>
#include <string>
#include <stdexcept>
#include <iomanip>
#include <thread>

#include "chrono/utils/ChUtils.h"
#include "chrono/utils/ChUtilsInputOutput.h"
#include "chrono/assets/ChVisualSystem.h"

#include "chrono_vehicle/ChVehicleModelData.h"
#include "chrono_vehicle/driver/ChPathFollowerDriver.h"
#include "chrono_vehicle/wheeled_vehicle/vehicle/WheeledVehicle.h"
#include "chrono_vehicle/wheeled_vehicle/tire/ChDeformableTire.h"
#include "chrono_vehicle/terrain/CRMTerrain.h"
#include "chrono_vehicle/utils/ChUtilsJSON.h"
#include "chrono_vehicle/utils/ChVehiclePath.h"
#include "chrono_vehicle/ChVehicleVisualSystem.h"

#include "chrono_thirdparty/filesystem/path.h"

#ifdef CHRONO_VSG
    #include "chrono_vehicle/wheeled_vehicle/ChWheeledVehicleVisualSystemVSG.h"
    #include "chrono_fsi/sph/visualization/ChFsiVisualizationVSG.h"
#endif

#ifdef CHRONO_POSTPROCESS
    #include "chrono_postprocess/ChGnuPlot.h"
#endif

#include "demos/SetChronoSolver.h"

using namespace chrono;
using namespace chrono::fsi;
using namespace chrono::fsi::sph;
using namespace chrono::vehicle;

using std::cout;
using std::cerr;
using std::cin;
using std::endl;

// ===================================================================================================================

// CRM terrain patch type
enum class PatchType { RECTANGULAR, MARKER_DATA, HEIGHT_MAP };
PatchType patch_type = PatchType::RECTANGULAR;

// Terrain dimensions (for RECTANGULAR or HEIGHT_MAP patch type)
double terrain_length = 20;
double terrain_width = 3;

// Vehicle specification files
std::string vehicle_json = "Polaris/Polaris.json";
std::string engine_json = "Polaris/Polaris_EngineSimpleMap.json";
std::string transmission_json = "Polaris/Polaris_AutomaticTransmissionSimpleMap.json";
std::string tire_json = "Polaris/Polaris_RigidTire.json";
////std::string tire_json = "Polaris/Polaris_ANCF4Tire_Lumped.json";

// Suspend vehicle
bool fix_chassis = false;

// ===================================================================================================================

std::shared_ptr<WheeledVehicle> CreateVehicle(const ChCoordsys<>& init_pos, bool& fea_tires);
void CreateFSIWheels(std::shared_ptr<WheeledVehicle> vehicle, CRMTerrain& terrain);
std::shared_ptr<ChBezierCurve> CreatePath(const std::string& path_file);

// ===================================================================================================================

int main(int argc, char* argv[]) {
    // ---------------- 
    // Problem settings
    // ----------------

    double target_speed = 7.0;
    double tend = 30;
    bool verbose = true;

    // Visualization settings
    bool render = true;                    // use run-time visualization
    double render_fps = 200;               // rendering FPS
    bool visualization_sph = true;         // render SPH particles
    bool visualization_bndry_bce = false;  // render boundary BCE markers
    bool visualization_rigid_bce = true;   // render wheel BCE markers

    // CRM material properties
    double density = 1700;
    double cohesion = 5e3;
    double friction = 0.8;
    double youngs_modulus = 1e6;
    double poisson_ratio = 0.3;

    // CRM (moving) active box dimension
    double active_box_dim = 0.8;
    double settling_time = 0;

    // Set SPH spacing
    double spacing = (patch_type == PatchType::MARKER_DATA) ? 0.02 : 0.04;

    // SPH integration sacheme
    IntegrationScheme integration_scheme = IntegrationScheme::RK2;

    // --------------
    // Create vehicle
    // --------------

    cout << "Create vehicle..." << endl;
    double vehicle_init_height = 0.25;
    bool fea_tires;
    auto vehicle = CreateVehicle(ChCoordsys<>(ChVector3d(3.5, 0, vehicle_init_height), QUNIT), fea_tires);
    vehicle->GetChassis()->SetFixed(fix_chassis);
    auto sysMBS = vehicle->GetSystem();

    // ---------------------------------
    // Set solver and integrator for MBD
    // ---------------------------------

    double step_size = 0;
    ChSolver::Type solver_type;
    ChTimestepper::Type integrator_type;

    if (fea_tires) {
        step_size = 1e-4;
        solver_type = ChSolver::Type::PARDISO_MKL;
        integrator_type = ChTimestepper::Type::EULER_IMPLICIT_LINEARIZED;
    } else {
        step_size = 5e-4;
        solver_type = ChSolver::Type::BARZILAIBORWEIN;
        integrator_type = ChTimestepper::Type::EULER_IMPLICIT_LINEARIZED;
    }

    int num_threads_chrono = std::min(8, ChOMP::GetNumProcs());
    int num_threads_collision = 1;
    int num_threads_eigen = 1;
    int num_threads_pardiso = std::min(8, ChOMP::GetNumProcs());

    SetChronoSolver(*sysMBS, solver_type, integrator_type, num_threads_pardiso);
    sysMBS->SetNumThreads(num_threads_chrono, num_threads_collision, num_threads_eigen);

    // Set collision system
    sysMBS->SetCollisionSystemType(ChCollisionSystem::Type::BULLET);

    // ----------------------
    // Create the CRM terrain
    // ----------------------

    CRMTerrain terrain(*sysMBS, spacing);
    ChFsiSystemSPH& sysFSI = terrain.GetSystemFSI();
    terrain.SetVerbose(verbose);
    terrain.SetGravitationalAcceleration(ChVector3d(0, 0, -9.81));
    terrain.SetStepSizeCFD(step_size);

    // Register the vehicle with the CRM terrain
    terrain.RegisterVehicle(vehicle.get());

    // Set SPH parameters and soil material properties
    ChFsiFluidSystemSPH::ElasticMaterialProperties mat_props;
    mat_props.density = density;
    mat_props.Young_modulus = youngs_modulus;
    mat_props.Poisson_ratio = poisson_ratio;
    mat_props.mu_I0 = 0.04;
    mat_props.mu_fric_s = friction;
    mat_props.mu_fric_2 = friction;
    mat_props.average_diam = 0.005;
    mat_props.cohesion_coeff = cohesion;
    terrain.SetElasticSPH(mat_props);

    // Set SPH solver parameters
    ChFsiFluidSystemSPH::SPHParameters sph_params;
    sph_params.integration_scheme = integration_scheme;
    sph_params.initial_spacing = spacing;
    sph_params.d0_multiplier = 1;
    sph_params.kernel_threshold = 0.8;
    sph_params.artificial_viscosity = 0.5;
    sph_params.consistent_gradient_discretization = false;
    sph_params.consistent_laplacian_discretization = false;
    sph_params.viscosity_type = ViscosityType::ARTIFICIAL_BILATERAL;
    sph_params.boundary_type = BoundaryType::ADAMI;
    terrain.SetSPHParameters(sph_params);

    // Set output level from SPH simulation
    terrain.SetOutputLevel(OutputLevel::STATE);

    // Add vehicle wheels as FSI solids
    CreateFSIWheels(vehicle, terrain);
    terrain.SetActiveDomain(ChVector3d(active_box_dim));
    terrain.SetActiveDomainDelay(settling_time);

    // Construct the terrain and associated path
    cout << "Create terrain..." << endl;
    std::shared_ptr<ChBezierCurve> path;
    switch (patch_type) {
        case PatchType::RECTANGULAR:
            // Create a rectangular terrain patch
            terrain.Construct({terrain_length, terrain_width, 0.25},  // length X width X height
                              ChVector3d(terrain_length / 2, 0, 0),   // patch center
                              BoxSide::ALL & ~BoxSide::Z_POS          // all boundaries, except top
            );
            // Create straight line path
            path = StraightLinePath(ChVector3d(0, 0, vehicle_init_height),
                                    ChVector3d(terrain_length, 0, vehicle_init_height), 1);
            break;
        case PatchType::HEIGHT_MAP:
            // Create a patch from a heigh field map image
            terrain.Construct(vehicle::GetDataFile("terrain/height_maps/bump64.bmp"),  // height map image file
                              terrain_length, terrain_width,                           // length (X) and width (Y)
                              {0, 0.3},                                                // height range
                              0.25,                                                    // depth
                              true,                                                    // uniform depth
                              ChVector3d(terrain_length / 2, 0, 0),                    // patch center
                              BoxSide::Z_NEG                                           // bottom wall
            );
            // Create straight line path
            path = StraightLinePath(ChVector3d(0, 0, vehicle_init_height),
                                    ChVector3d(terrain_length, 0, vehicle_init_height), 1);
            break;
        case PatchType::MARKER_DATA:
            // Create a patch using SPH particles and BCE markers from files
            terrain.Construct(vehicle::GetDataFile("terrain/sph/S-lane_RMS/sph_particles.txt"),  // SPH marker locations
                              vehicle::GetDataFile("terrain/sph/S-lane_RMS/bce_markers.txt"),    // BCE marker locations
                              VNULL);
            // Create path from data file
            path = CreatePath("terrain/sph/S-lane_RMS/path.txt");
            break;
    }

    // Initialize the terrain system
    terrain.Initialize();

    auto aabb = terrain.GetSPHBoundingBox();
    cout << "  SPH particles:     " << terrain.GetNumSPHParticles() << endl;
    cout << "  Bndry BCE markers: " << terrain.GetNumBoundaryBCEMarkers() << endl;
    cout << "  SPH AABB:          " << aabb.min << "   " << aabb.max << endl;

    // Set maximum vehicle X location (based on CRM patch size)
    double x_max = aabb.max.x() - 4.5;

    // --------------------------------
    // Create the path-following driver
    // --------------------------------

    cout << "Create path..." << endl;
    ChPathFollowerDriver driver(*vehicle, path, "my_path", target_speed);
    driver.GetSteeringController().SetLookAheadDistance(2.0);
    driver.GetSteeringController().SetGains(1.0, 0, 0);
    driver.GetSpeedController().SetGains(0.6, 0.05, 0);
    driver.Initialize();

    // -----------------------------
    // Set up output
    // -----------------------------

    std::string out_dir = GetChronoOutputPath() + "CRM_Wheeled_Vehicle/";
    if (!filesystem::create_directory(filesystem::path(out_dir))) {
        cerr << "Error creating directory " << out_dir << endl;
        return 1;
    }

    std::string out_file = out_dir + "/results.txt";
    utils::ChWriterCSV csv(" ");

    // -----------------------------
    // Create run-time visualization
    // -----------------------------

    std::shared_ptr<ChVehicleVisualSystem> vis;

#ifdef CHRONO_VSG
    if (render) {
        // FSI plugin
        auto visFSI = chrono_types::make_shared<ChFsiVisualizationVSG>(&sysFSI);
        visFSI->EnableFluidMarkers(visualization_sph);
        visFSI->EnableBoundaryMarkers(visualization_bndry_bce);
        visFSI->EnableRigidBodyMarkers(visualization_rigid_bce);
        visFSI->SetSPHColorCallback(chrono_types::make_shared<ParticleHeightColorCallback>(ChColor(0.10f, 0.40f, 0.65f), aabb.min.z(), aabb.max.z()));

        // Wheeled vehicle VSG visual system (attach visFSI as plugin)
        auto visVSG = chrono_types::make_shared<ChWheeledVehicleVisualSystemVSG>();
        visVSG->AttachVehicle(vehicle.get());
        visVSG->AttachPlugin(visFSI);
        visVSG->SetWindowTitle("Wheeled vehicle on CRM deformable terrain");
        visVSG->SetWindowSize(1280, 800);
        visVSG->SetWindowPosition(100, 100);
        visVSG->EnableSkyBox();
        visVSG->SetLightIntensity(1.0f);
        visVSG->SetLightDirection(1.5 * CH_PI_2, CH_PI_4);
        visVSG->SetCameraAngleDeg(40);
        visVSG->SetChaseCamera(VNULL, 6.0, 2.0);
        visVSG->SetChaseCameraPosition(ChVector3d(0, 8, 1.5));

        visVSG->Initialize();
        vis = visVSG;
    }
#else
    render = false;
#endif

    // ---------------
    // Simulation loop
    // ---------------

    double time = 0;
    int sim_frame = 0;
    int render_frame = 0;
    bool braking = false;

    cout << "Start simulation..." << endl;

    while (time < tend) {
        const auto& veh_loc = vehicle->GetPos();

        // Set current driver inputs
        auto driver_inputs = driver.GetInputs();

        // Ramp up throttle to value requested by the cruise controller
        if (time < 0.5) {
            driver_inputs.m_throttle = 0;
            driver_inputs.m_braking = 1;
        } else {
            ChClampValue(driver_inputs.m_throttle, driver_inputs.m_throttle, (time - 0.5) / 0.5);
        }

        // Stop vehicle before reaching end of terrain patch, then end simulation after 2 more second2
        if (veh_loc.x() > x_max) {
            driver_inputs.m_throttle = 0;
            driver_inputs.m_braking = 1;
            if (!braking) {
                cout << "Start braking..." << endl;
                tend = time + 2;
                braking = true;
            }
        }

        // Run-time visualization
        if (render && time >= render_frame / render_fps) {
            if (!vis->Run())
                break;
            vis->Render();
            render_frame++;
        }
        if (!render) {
            std::cout << time << "  " << terrain.GetRtfCFD() << "  " << terrain.GetRtfMBD() << std::endl;
        }

        // Synchronize systems
        driver.Synchronize(time);
        vis->Synchronize(time, driver_inputs);
        terrain.Synchronize(time);
        vehicle->Synchronize(time, driver_inputs, terrain);

        // Advance system state
        driver.Advance(step_size);
        vis->Advance(step_size);
        // Coupled FSI problem (CRM terrain + vehicle)
        terrain.Advance(step_size);

        csv << time << vehicle->GetPos() << vehicle->GetSpeed() << endl;

        time += step_size;
        sim_frame++;
    }

    csv.WriteToFile(out_file);

#ifdef CHRONO_POSTPROCESS
    postprocess::ChGnuPlot gplot(out_dir + "/height.gpl");
    gplot.SetGrid();
    std::string title = "Vehicle ref frame height";
    gplot.SetTitle(title);
    gplot.SetLabelX("time (s)");
    gplot.SetLabelY("height (m)");
    gplot.Plot(out_file, 1, 4, "", " with lines lt -1 lw 2 lc rgb'#3333BB' ");
#endif

    return 0;
}

// ===================================================================================================================

std::shared_ptr<WheeledVehicle> CreateVehicle(const ChCoordsys<>& init_pos, bool& fea_tires) {
    fea_tires = false;

    // Create and initialize the vehicle
    auto vehicle = chrono_types::make_shared<WheeledVehicle>(vehicle::GetDataFile(vehicle_json), ChContactMethod::SMC);
    vehicle->Initialize(init_pos);
    vehicle->GetChassis()->SetFixed(false);
    vehicle->SetChassisVisualizationType(VisualizationType::MESH);
    vehicle->SetSuspensionVisualizationType(VisualizationType::PRIMITIVES);
    vehicle->SetSteeringVisualizationType(VisualizationType::PRIMITIVES);
    vehicle->SetWheelVisualizationType(VisualizationType::MESH);

    // Create and initialize the powertrain system
    auto engine = ReadEngineJSON(vehicle::GetDataFile(engine_json));
    auto transmission = ReadTransmissionJSON(vehicle::GetDataFile(transmission_json));
    auto powertrain = chrono_types::make_shared<ChPowertrainAssembly>(engine, transmission);
    vehicle->InitializePowertrain(powertrain);

    // Create and initialize the tires
    for (auto& axle : vehicle->GetAxles()) {
        for (auto& wheel : axle->GetWheels()) {
            auto tire = ReadTireJSON(vehicle::GetDataFile(tire_json));
            vehicle->InitializeTire(tire, wheel, VisualizationType::MESH);
            if (std::dynamic_pointer_cast<ChDeformableTire>(tire))
                fea_tires = true;
        }
    }

    return vehicle;
}

void CreateFSIWheels(std::shared_ptr<WheeledVehicle> vehicle, CRMTerrain& terrain) {
    std::string mesh_filename = vehicle::GetDataFile("Polaris/meshes/Polaris_tire_collision.obj");
    utils::ChBodyGeometry geometry;
    geometry.materials.push_back(ChContactMaterialData());
    geometry.coll_meshes.push_back(utils::ChBodyGeometry::TrimeshShape(VNULL, mesh_filename, VNULL));

    for (auto& axle : vehicle->GetAxles()) {
        for (auto& wheel : axle->GetWheels()) {
            auto tire_fea = std::dynamic_pointer_cast<ChDeformableTire>(wheel->GetTire());
            if (tire_fea) {
                auto mesh = tire_fea->GetMesh();
                if (mesh->GetNumContactSurfaces() > 0) {
                    auto surf = mesh->GetContactSurface(0);
                    cout << "FEA tire HAS contact surface: ";
                    if (std::dynamic_pointer_cast<fea::ChContactSurfaceNodeCloud>(surf))
                        cout << " NODE_CLOUD" << endl;
                    else
                        cout << " TRI_MESH" << endl;
                } else {
                    cout << "FEA tire DOES NOT HAVE contact surface!" << endl;
                }
                terrain.AddFeaMesh(mesh, false);
            } else {
                terrain.AddRigidBody(wheel->GetSpindle(), geometry, false);
            }
        }
    }
}

std::shared_ptr<ChBezierCurve> CreatePath(const std::string& path_file) {
    // Open input file
    std::ifstream ifile(vehicle::GetDataFile(path_file));
    std::string line;

    // Read number of knots and type of curve
    size_t numPoints;
    size_t numCols;

    std::getline(ifile, line);
    std::istringstream iss(line);
    iss >> numPoints >> numCols;

    assert(numCols == 3);

    // Read path points
    std::vector<ChVector3d> points;

    for (size_t i = 0; i < numPoints; i++) {
        double x, y, z;
        std::getline(ifile, line);
        std::istringstream jss(line);
        jss >> x >> y >> z;
        points.push_back(ChVector3d(x, y, z));
    }

    // Include point beyond CRM patch
    {
        auto np = points.size();
        points.push_back(2.0 * points[np - 1] - points[np - 2]);
    }

    // Raise all path points
    for (auto& p : points)
        p.z() += 0.1;

    ifile.close();

    return std::shared_ptr<ChBezierCurve>(new ChBezierCurve(points));
}

File: src/demos/python/vehicle/demo_VEH_CRMTerrain_WheeledVehicle.py
# =============================================================================
# PROJECT CHRONO - http://projectchrono.org
#
# Copyright (c) 2014 projectchrono.org
# All rights reserved.
#
# Use of this source code is governed by a BSD-style license that can be found
# in the LICENSE file at the top level of the distribution and at
# http://projectchrono.org/license-chrono.txt.
#
# =============================================================================
# Authors: Huzaifa Unjhawala
# =============================================================================
#
# Demonstration of Polaris RZR accelerating on a patch of CRM Terrain
#
#
# =============================================================================

import pychrono.core as chrono
import pychrono.vehicle as veh
import pychrono.fsi as fsi
import pychrono.vsg3d as vsg
import os

def CreateFSIWheels(vehicle, terrain):
    """
    Add vehicle wheels as FSI solids to the CRM terrain.
    
    Args:
        vehicle: WheeledVehicle object
        terrain: CRMTerrain object
    """
    mesh_filename = veh.GetVehicleDataFile("Polaris/meshes/Polaris_tire_collision.obj")
    
    # Create geometry for rigid wheels
    geometry = chrono.ChBodyGeometry()
    geometry.coll_meshes.append(chrono.TrimeshShape(chrono.VNULL, chrono.QUNIT, mesh_filename, chrono.VNULL))
    # Iterate through all axles and wheels
    for axle in vehicle.GetAxles():
        for wheel in axle.GetWheels():
            tire = wheel.GetTire()
            # Check if this is a deformable tire (FEA)
            try:
                # Try to get FEA mesh if it's a deformable tire
                if hasattr(tire, 'GetMesh'):
                    mesh = tire.GetMesh()
                    terrain.AddFeaMesh(mesh, False)
                else:
                    # Rigid tire - add as rigid body
                    terrain.AddRigidBody(wheel.GetSpindle(), geometry, False)
            except Exception as e:
                print(f"Error processing wheel: {e}")
                # If we can't access FEA mesh methods, treat as rigid tire
                terrain.AddRigidBody(wheel.GetSpindle(), geometry, False)


# Set output root directory
chrono.SetChronoOutputPath("../DEMO_OUTPUT/")

# Problem settings (mirroring the C++ demo)
target_speed = 7.0
tend = 30.0
verbose = True

# Visualization settings
render = True
render_fps = 200
visualization_sph = True
visualization_bndry_bce = False
visualization_rigid_bce = False

# CRM material properties
density = 1700
cohesion = 5e3
friction = 0.8
youngs_modulus = 1e6
poisson_ratio = 0.3

# CRM (moving) active box dimension
active_box_dim = 0.8
settling_time = 0

# Set SPH spacing
spacing = 0.04

# Vehicle specification files (adjust paths as needed)
vehicle_json = "Polaris/Polaris.json"
engine_json = "Polaris/Polaris_EngineSimpleMap.json"
transmission_json = "Polaris/Polaris_AutomaticTransmissionSimpleMap.json"
tire_json = "Polaris/Polaris_RigidTire.json"
# tire_json = "Polaris/Polaris_ANCF4Tire_Lumped.json"
if(tire_json.find("ANCF4Tire") != -1):
    fea_tires = True
else:
    fea_tires = False

# Visualization type for vehicle parts (PRIMITIVES, MESH, or NONE)
chassis_vis_type = chrono.VisualizationType_MESH
suspension_vis_type = chrono.VisualizationType_PRIMITIVES
steering_vis_type = chrono.VisualizationType_PRIMITIVES
wheel_vis_type = chrono.VisualizationType_MESH
tire_vis_type = chrono.VisualizationType_MESH

# --------------
# Create vehicle
# --------------
print("Create vehicle...")
vehicle_init_height = 0.25
# This assumes a WheeledVehicle class and similar API as C++
vehicle = veh.WheeledVehicle(veh.GetVehicleDataFile(vehicle_json), chrono.ChContactMethod_SMC)
vehicle.Initialize(chrono.ChCoordsysd(chrono.ChVector3d(3.5, 0, vehicle_init_height), chrono.QUNIT))
vehicle.GetChassis().SetFixed(False)
vehicle.SetChassisVisualizationType(chassis_vis_type)
vehicle.SetSuspensionVisualizationType(suspension_vis_type)
vehicle.SetSteeringVisualizationType(steering_vis_type)
vehicle.SetWheelVisualizationType(wheel_vis_type)
vehicle.SetTireVisualizationType(tire_vis_type)

# Powertrain
engine = veh.ReadEngineJSON(veh.GetVehicleDataFile(engine_json))
transmission = veh.ReadTransmissionJSON(veh.GetVehicleDataFile(transmission_json))
powertrain = veh.ChPowertrainAssembly(engine, transmission)
vehicle.InitializePowertrain(powertrain)

# Tires
for axle in vehicle.GetAxles():
    for wheel in axle.GetWheels():
        tire = veh.ReadTireJSON(veh.GetVehicleDataFile(tire_json))
        vehicle.InitializeTire(tire, wheel, chrono.VisualizationType_MESH)

        

sysMBS = vehicle.GetSystem()

# Set solver and integrator based on tire type
if fea_tires:
    import pychrono.pardisomkl as mkl
    step_size = 1e-4
    # solver_type = mkl.ChSolverPardisoMKL
    integrator_type = chrono.ChTimestepper.Type_EULER_IMPLICIT_LINEARIZED
    # Set number of threads
    num_threads_chrono = 8
    num_threads_collision = 1
    num_threads_eigen = 1
    num_threads_pardiso = 8

    # Set solver and integrator
    sysMBS.SetSolver(mkl.ChSolverPardisoMKL(num_threads_pardiso))
    sysMBS.SetTimestepperType(integrator_type)
    sysMBS.SetNumThreads(num_threads_chrono, num_threads_collision, num_threads_eigen)
else:
    step_size = 5e-4
    solver_type = chrono.ChSolver.Type_BARZILAIBORWEIN
    integrator_type = chrono.ChTimestepper.Type_EULER_IMPLICIT_LINEARIZED
    # Set number of threads
    num_threads_chrono = 8
    num_threads_collision = 1
    num_threads_eigen = 6
    num_threads_pardiso = 0

    # Set solver and integrator
    sysMBS.SetSolverType(solver_type)
    sysMBS.SetTimestepperType(integrator_type)
    sysMBS.SetNumThreads(num_threads_chrono, num_threads_collision, num_threads_eigen)


# Set collision system
sysMBS.SetCollisionSystemType(chrono.ChCollisionSystem.Type_BULLET)

# ----------------------
# Create the CRM terrain
# ----------------------
# Use the proper CRMTerrain class from vehicle module
terrain = veh.CRMTerrain(sysMBS, spacing)
sysFSI = terrain.GetFsiSystemSPH()
terrain.SetVerbose(verbose)
terrain.SetGravitationalAcceleration(chrono.ChVector3d(0, 0, -9.81))
terrain.SetStepSizeCFD(step_size)

# Register the vehicle with the CRM terrain
terrain.RegisterVehicle(vehicle)

# Set SPH parameters and soil material properties
mat_props = fsi.ElasticMaterialProperties()
mat_props.density = density
mat_props.Young_modulus = youngs_modulus
mat_props.Poisson_ratio = poisson_ratio
mat_props.mu_I0 = 0.04
mat_props.mu_fric_s = friction
mat_props.mu_fric_2 = friction
mat_props.average_diam = 0.005
mat_props.cohesion_coeff = cohesion
terrain.SetElasticSPH(mat_props)

# Set SPH solver parameters
sph_params = fsi.SPHParameters()
sph_params.integration_scheme = fsi.IntegrationScheme_RK2
sph_params.initial_spacing = spacing
sph_params.d0_multiplier = 1.2
sph_params.kernel_threshold = 0.8
sph_params.artificial_viscosity = 0.5
sph_params.shifting_method = fsi.ShiftingMethod_PPST
sph_params.shifting_ppst_push = 3.0
sph_params.shifting_ppst_pull = 1.0
sph_params.consistent_gradient_discretization = False
sph_params.consistent_laplacian_discretization = False
sph_params.viscosity_method = fsi.ViscosityMethod_ARTIFICIAL_BILATERAL
sph_params.boundary_method = fsi.BoundaryMethod_ADAMI
terrain.SetSPHParameters(sph_params)

# Set output level from SPH simulation
# terrain.SetOutputLevel(fsi.OutputLevel_STATE)

# Add vehicle wheels as FSI solids
print("Adding vehicle wheels as FSI solids...")
CreateFSIWheels(vehicle, terrain)

# CRITICAL: Use correct active domain setup
terrain.SetActiveDomain(chrono.ChVector3d(active_box_dim))
terrain.SetActiveDomainDelay(settling_time)

# Construct the terrain and associated path
print("Create terrain...")
terrain_length = 20
terrain_width = 3
terrain.Construct(chrono.ChVector3d(terrain_length, terrain_width, 0.25), 
                  chrono.ChVector3d(terrain_length / 2, 0, 0), 
                  (fsi.BoxSide_ALL & ~fsi.BoxSide_Z_POS))

# Create straight line path
path = veh.StraightLinePath(chrono.ChVector3d(0, 0, vehicle_init_height), 
                            chrono.ChVector3d(terrain_length, 0, vehicle_init_height), 1)

# Initialize the terrain system
terrain.Initialize()

aabb = terrain.GetSPHBoundingBox()
print(f"  SPH particles:     {terrain.GetNumSPHParticles()}")
print(f"  Bndry BCE markers: {terrain.GetNumBoundaryBCEMarkers()}")
print(f"  SPH AABB:          {aabb.min}   {aabb.max}")

# Set maximum vehicle X location (based on CRM patch size)
x_max = aabb.max.x - 4.5

# --------------------------------
# Create the path-following driver
# --------------------------------
print("Create path...")
driver = veh.ChPathFollowerDriver(vehicle, path, "my_path", target_speed)
driver.GetSteeringController().SetLookAheadDistance(2.0)
driver.GetSteeringController().SetGains(1.0, 0, 0)
driver.GetSpeedController().SetGains(0.6, 0.05, 0)
driver.Initialize()

# -----------------------------
# Set up output
# -----------------------------
out_dir = chrono.GetChronoOutputPath() + "CRM_Wheeled_Vehicle/"
os.makedirs(out_dir, exist_ok=True)
out_file = os.path.join(out_dir, "results.txt")

# Use a simple CSV writer
import csv
csvfile = open(out_file, 'w', newline='')
csvwriter = csv.writer(csvfile, delimiter=' ')

# -----------------------------
# Create run-time visualization
# -----------------------------
vis = None
if render:
    # FSI plugin
    col_callback = fsi.ParticleHeightColorCallback(aabb.min.z, aabb.max.z)
    visFSI = fsi.ChSphVisualizationVSG(sysFSI)
    visFSI.EnableFluidMarkers(visualization_sph)
    visFSI.EnableBoundaryMarkers(visualization_bndry_bce)
    visFSI.EnableRigidBodyMarkers(visualization_rigid_bce)
    visFSI.SetSPHColorCallback(col_callback, chrono.ChColormap.Type_BROWN)

    visVSG = veh.ChWheeledVehicleVisualSystemVSG()
    visVSG.AttachVehicle(vehicle)
    visVSG.AttachPlugin(visFSI)
    visVSG.SetWindowTitle("Wheeled vehicle on CRM deformable terrain")
    visVSG.SetWindowSize(1280, 800)
    visVSG.SetWindowPosition(100, 100)
    visVSG.EnableSkyBox()
    visVSG.SetLightIntensity(1.0)
    visVSG.SetLightDirection(1.5 * chrono.CH_PI_2, chrono.CH_PI_4)
    visVSG.SetCameraAngleDeg(40)
    visVSG.SetChaseCamera(chrono.VNULL, 6.0, 2.0)
    visVSG.SetChaseCameraPosition(chrono.ChVector3d(0, 8, 1.5))
    visVSG.Initialize()
    vis = visVSG

# ---------------
# Simulation loop
# ---------------
time = 0
sim_frame = 0
render_frame = 0
braking = False

print("Start simulation...")
while time < tend:
    veh_loc = vehicle.GetPos()
    driver_inputs = driver.GetInputs()

    # Ramp up throttle to value requested by the cruise controller
    if time < 0.5:
        driver_inputs.m_throttle = 0
        driver_inputs.m_braking = 1
    else:
        driver_inputs.m_throttle = min(driver_inputs.m_throttle, (time - 0.5) / 0.5)

    # Stop vehicle before reaching end of terrain patch, then end simulation after 2 more seconds
    if veh_loc.x > x_max:
        driver_inputs.m_throttle = 0
        driver_inputs.m_braking = 1
        if not braking:
            print("Start braking...")
            tend = time + 2
            braking = True

    # Run-time visualization
    if render and time >= render_frame / render_fps:
        if not vis.Run():
            break
        vis.Render()
        render_frame += 1

    # Synchronize systems
    driver.Synchronize(time)
    if vis:
        vis.Synchronize(time, driver_inputs)
    terrain.Synchronize(time)
    vehicle.Synchronize(time, driver_inputs, terrain)

    # Advance system state
    driver.Advance(step_size)
    if vis:
        vis.Advance(step_size)
        vis.WriteImageToFile(os.path.join(out_dir, f"img_{render_frame + 1:05d}.bmp"))
    terrain.Advance(step_size)

    # Output results
    csvwriter.writerow([time, veh_loc.x, veh_loc.y, veh_loc.z, vehicle.GetSpeed()])

    time += step_size
    sim_frame += 1

csvfile.close()
print("Simulation complete. Results written to:", out_file)

